apps/frontend/src/lib/components/ChatMessage.svelte

```diff
@@ -26,21 +52,100 @@
+				onmouseleave={() => (isHovered = false)}
+				class="relative pb-10"
+			>
+				{#if message.type === 'ai'}
+					<Card
+						class="w-full max-w-none border border-gray-200 bg-gray-50 p-4 text-sm shadow-sm dark:border-gray-600 dark:bg-gray-800"
+					>
+						<div class="prose prose-gray dark:prose-invert max-w-none leading-relaxed">
+							<Markdown md={message.text} {plugins} />
+						</div>
+					</Card>
+
+					<!-- AI Message Actions -->
+					<div
+						class="absolute bottom-2 left-0 flex items-center gap-2 transition-all duration-300 ease-in-out"
+						style="opacity: {isHovered ? '1' : '0'}; transform: translateY({isHovered
+							? '0'
+							: '-4px'});"
+					>
+						<Button
+							onclick={handleCopy}
+							class="p-1.5!"
+							color="alternative"
+							size="xs"
+							title={m.message_copy()}
+							disabled
+						>
+							<ClipboardOutline size="xs" />
+						</Button>
+						<Button
+							onclick={() => onRegenerate?.(message)}
+							class="p-1.5!"
+							color="alternative"
+							size="xs"
+							title={m.message_regenerate()}
+							disabled
+						>
+							<ArrowsRepeatOutline size="xs" />
+						</Button>
+						<div class="ml-2 flex gap-1 border-l border-gray-300 pl-2 dark:border-gray-600">
+							<Button
+								onclick={() => handleFeedback('up')}
+								class="p-1.5! {feedbackGiven === 'up' ? 'bg-gray-200 dark:bg-gray-700' : ''}"
+								color="alternative"
+								size="xs"
+								title={m.message_feedback_good()}
+								disabled
+							>
+								<ThumbsUpOutline size="xs" />
+							</Button>
+							<Button
+								onclick={() => handleFeedback('down')}
+								class="p-1.5! {feedbackGiven === 'down' ? 'bg-gray-200 dark:bg-gray-700' : ''}"
+								color="alternative"
+								size="xs"
+								title={m.message_feedback_bad()}
+								disabled
+							>
+								<ThumbsDownOutline size="xs" />
+							</Button>
+						</div>
 					</div>
-				</Card>
-			{:else}
-				<Card class="w-full max-w-none border-0 bg-gray-800 p-4 text-sm shadow-sm dark:bg-gray-700">
-					<div class="prose prose-invert max-w-none leading-relaxed whitespace-pre-wrap">
-						{message.text}
+				{:else}
+					<Card
+						class="w-full max-w-none border-0 bg-gray-800 p-4 text-sm shadow-sm dark:bg-gray-700"
+					>
+						<div class="prose prose-invert max-w-none leading-relaxed whitespace-pre-wrap">
+							{message.text}
+						</div>
+					</Card>
+
+					<!-- User Message Actions -->
+					<div
+						class="absolute right-0 bottom-2 flex items-center gap-2 transition-all duration-300 ease-in-out"
+						style="opacity: {isHovered ? '1' : '0'}; transform: translateY({isHovered
+							? '0'
+							: '-4px'});"
+					>
+						<Button
+							onclick={() => onEdit?.(message)}
+							class="p-1.5!"
+							color="alternative"
+							size="xs"
+							title={m.message_edit()}
+							disabled
+						>
+							<PenOutline size="xs" />
+						</Button>
```

Comment:
This is enough code that I would suggest using one or two custom components. ChatUserFeedback/ChatAIFeedback or  a single ChatFeedback (depending on how dry we wanna go).

The idea is to keep our code maximally semantic/legible.